front/lib/api/assistant/actions/websearch.ts

```diff
@@ -1,10 +1,95 @@
-import type { AgentActionSpecification, Result } from "@dust-tt/types";
-import { Ok } from "@dust-tt/types";
-import type { WebsearchConfigurationType } from "@dust-tt/types/dist/front/assistant/actions/websearch";
+import type {
+  AgentActionSpecification,
+  FunctionCallType,
+  FunctionMessageTypeModel,
+  ModelId,
+  ModelMessageType,
+  Result,
+  WebsearchActionOutputType,
+  WebsearchConfigurationType,
+  WebsearchErrorEvent,
+  WebsearchParamsEvent,
+  WebsearchSuccessEvent,
+} from "@dust-tt/types";
+import {
+  BaseAction,
+  cloneBaseConfig,
+  DustProdActionRegistry,
+  Ok,
+} from "@dust-tt/types";
 
+import { runActionStreamed } from "@app/lib/actions/server";
 import type { BaseActionRunParams } from "@app/lib/api/assistant/actions/types";
 import { BaseActionConfigurationServerRunner } from "@app/lib/api/assistant/actions/types";
 import type { Authenticator } from "@app/lib/auth";
+import { AgentWebsearchAction } from "@app/lib/models/assistant/actions/websearch";
+import logger from "@app/logger/logger";
+
+interface WebsearchActionBlob {
+  id: ModelId; // AgentWebsearchAction
+  agentMessageId: ModelId;
+  query: string;
+  output: WebsearchActionOutputType | null;
+  functionCallId: string | null;
+  functionCallName: string | null;
+  step: number;
+}
+
+export class WebsearchAction extends BaseAction {
+  readonly agentMessageId: ModelId;
+  readonly query: string;
+  readonly output: WebsearchActionOutputType | null;
+  readonly functionCallId: string | null;
+  readonly functionCallName: string | null;
+  readonly step: number;
+
+  constructor(blob: WebsearchActionBlob) {
+    super(blob.id, "websearch_action");
+
+    this.agentMessageId = blob.agentMessageId;
+    this.query = blob.query;
+    this.output = blob.output;
+    this.functionCallId = blob.functionCallId;
+    this.functionCallName = blob.functionCallName;
+    this.step = blob.step;
+  }
+
+  renderForModel(): ModelMessageType {
+    let content = "";
+
+    // Note action.output can be any valid JSON including null.
+    content += `WEBSEARCH OUTPUT:\n`;
+    content += `${JSON.stringify(this.output, null, 2)}\n`;
```

Comment:
I think we can have a special case if this.output is null saying (web search failed) ?